From 5f7bd9612edd5ab6b8cd45148be75b3fd7c86cf8 Mon Sep 17 00:00:00 2001
From: Sumit Bose <sbose@redhat.com>
Date: Wed, 14 Jan 2026 13:58:41 +0100
Subject: [PATCH] ipa-otpd: do not pass reference of data_out->jdata

The 'o' format specifier will pass the reference of data_out->jdata to
the created object and data_out->jdata will be dereferenced if the new
object is dereferenced. Doing an explicit dereferenced of
data_out->jdata later on will cause a double-free crash.

With 'O' the reference counter of data_out->jdata is just incremented
and so it can be properly dereferenced.

Fixes: https://pagure.io/freeipa/issue/9906
Signed-off-by: Sumit Bose <sbose@redhat.com>
Reviewed-By: Alexander Bokovoy <abokovoy@redhat.com>
Reviewed-By: Rob Crittenden <rcritten@redhat.com>
Reviewed-By: Rob Crittenden <rcritten@redhat.com>
Reviewed-By: Alexander Bokovoy <abokovoy@redhat.com>
---
 daemons/ipa-otpd/passkey.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/daemons/ipa-otpd/passkey.c b/daemons/ipa-otpd/passkey.c
index ad3c45467..51d5c316c 100644
--- a/daemons/ipa-otpd/passkey.c
+++ b/daemons/ipa-otpd/passkey.c
@@ -437,7 +437,9 @@ static int prepare_rad_reply(struct otpd_queue_item *item)
         goto done;
     }
 
-    jtmp = json_pack("{s:i, s:s, s:o}", "phase", item->passkey->data_out->phase,
+    /* item->passkey->data_out->jdata will be dereferenced later, so we have
+     * to use 'O' to keep the reference. */
+    jtmp = json_pack("{s:i, s:s, s:O}", "phase", item->passkey->data_out->phase,
                                         "state", item->passkey->data_out->state,
                                         "data", item->passkey->data_out->jdata);
     if (jtmp == NULL) {
-- 
2.50.1

